RavenDB-7513 : Support date operations

aviv86 opened the pull request.

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -113,8 +113,35 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                            }
+                        }
+
+                        writer.Write($"new Date({argsStr})");
```
Are you sure that the position & number of arguments match between them?

aviv86 replied:
yes, as long as it's not one of the DateTime() overloads that receives a Calender and/or DateTimeKind arguments.

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -124,19 +151,54 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
                 using (javascriptWriter.Operation(node))
                 {
-                    if (node.Expression == Parameter)
+                    if (node.Type == typeof(DateTime))
```
Please add comments here showing the C# you are matching and the JS that is being generated.

ayende commented on src/Raven.Server/Documents/Patch/DocumentPatcherBase.cs:
```diff
@@ -129,6 +129,10 @@ private void PrepareEngine(PatchRequest patch, PatcherOperationScope scope, Engi
 
             CustomizeEngine(jintEngine, scope);
 
+            jintEngine.SetValue("$now", JsValue.FromObject(jintEngine, DateTime.Now));
```
It would be better to make these into functions, not static values.
A major issue here is that we are caching the engine, so that can't be made to work with statically cached values.

maximburyak replied:
I think that we better keep it as a value, especially because of the Jint (or any other) costs of calling a funciton. The "prepare engine" funciton is called anyway, even in the new implementation using the 'Initalize" method

ayende commented on src/Raven.Server/Documents/Patch/DocumentPatcherBase.cs:
```diff
@@ -129,6 +129,10 @@ private void PrepareEngine(PatchRequest patch, PatcherOperationScope scope, Engi
 
+            jintEngine.SetValue("$now", JsValue.FromObject(jintEngine, DateTime.Now));
+            jintEngine.SetValue("$today", JsValue.FromObject(jintEngine, DateTime.Today));
+            jintEngine.SetValue("$utcnow", JsValue.FromObject(jintEngine, DateTime.Now - (DateTime.Now - DateTime.UtcNow)));
```
Why this computation, you can just use `DateTime.UtcNow`

aviv86 replied:
I tried it first and it didn't work, the value was DateTime.Now

ayende commented on test/FastTests/Client/Subscriptions/LinqMethodsSupportForSubscriptionScript.cs:
```diff
@@ -517,6 +517,315 @@ await session.StoreAsync(new SupportCall()
+                var options = new SubscriptionCreationOptions<SupportCall>
+                {
+                    Criteria = new SubscriptionCriteria<SupportCall>(
+                        call => call.Started > DateTime.Today)
```
This works, but I'm wondering how useful that would be.
@maximburyak Comment on actual usage?

maximburyak replied:
About actual usage: that's nice and readable and very convenient for basic usage.
But
I think that now that we are pushing much more javascript to our api, maybe we should expose it more, rather then hiding it. It's important both for "educational" and api consistency reasons. Moreover, we've seen in the past that duality between "basic" and "pro" APIs hides the true potential of tools from our users.

aviv86 committed:
RavenDB-7513 : fix PR comments

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -113,8 +113,36 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                if (newExp != null && newExp.Type == typeof(DateTime))
+                {
+                    var args = newExp.Arguments;
+                    var argsStr = string.Empty;
```
Can you build it inline? If we call it each time, that can cause a lot of allocations

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -124,19 +152,60 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                        //match DateTime expressions like call.Started, user.DateOfBirth, etc
+                        if (node.Expression == Parameter)
+                        {
+                            //transle it to Date.parse(this.Started)
```
Typo in comment

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -124,19 +152,60 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                        switch (node.Member.Name)
+                        {
+                            case "Now":
+                                javascriptWriter.Write("Date.now()");
```
needs to be `new Date()` here.

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -124,19 +152,60 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
+                                javascriptWriter.Write("Date.now()");
+                                break;
+                            case "UtcNow":
+                                var utc = DateTime.UtcNow;
```
No, you are recording the time you are doing this, you need to record the time you actually invoked it.

aviv86 committed:
RavenDB-7513 : fix more PR comments

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -232,11 +219,12 @@ private static bool IsCollection(Type type)
 
         public SubscriptionCriteria(Expression<Func<T, bool>> predicate)
         {
+            var utc = DateTime.UtcNow;
```
Cannot do that.

ayende commented on src/Raven.Client/Documents/Subscriptions/SubscriptionCriteria.cs:
```diff
@@ -119,25 +120,12 @@ public override void ConvertToJavascript(JavascriptConversionContext context)
                     using (writer.Operation(newExp))
                     {
-                        writer.Write($"new Date({argsStr})");
+                        writer.Write($"new Date({string.Join(", ", newExp.Arguments)})");
```
Still allocating a lot

aviv86 committed:
RavenDB-7513:
- fix handling new DateTime expressions
- fix handling UtcNow
- move all Lambda2Js extensions to a separate class
- rename tests class
